crates/nostr-relay-builder/src/local/inner.rs

```diff
@@ -766,27 +769,39 @@ impl InnerLocalRelay {
 
             // Check mode and if it's authenticated
             if nip42.mode.is_read() && !session.nip42.is_authenticated() {
-                // Generate and send AUTH challenge
-                send_msg(
+                return send_auth_and_close(
                     ws_tx,
-                    RelayMessage::Auth {
-                        challenge: Cow::Owned(session.nip42.generate_challenge()),
-                    },
+                    subscription_id,
+                    session.nip42.generate_challenge(),
                 )
-                .await?;
+                .await;
+            }
+        }
 
-                // Return error
-                return send_msg(
+        // Check if NIP-42 DMs are enabled and any filter includes the GiftWrap kind
+        let giftwraps_filters = find_filters_with_kind(&filters, &Kind::GiftWrap);
+        if self.auth_dm && !giftwraps_filters.is_empty() {
```

Comment:
Hey @yukibtc, what do you think about changing this to this

```rust
        let giftwraps_filters = self
            .auth_dm
            .then_some(find_filters_with_kind(&filters, &Kind::GiftWrap))
            .flatten();
        if let Some(giftwraps_filters) = giftwraps_filters {
```

So we don't have to collect the `giftwraps_filters` if `self.auth_dm` is `false`.

---

I also think this is better, but I think it's not **simple**, let me know.

```rust
        if let Some(giftwraps_filters) = self
            .auth_dm
            .then_some(find_filters_with_kind(&filters, &Kind::GiftWrap))
            .flatten()
        {
```